src/Compilers/Core/Portable/Diagnostic/LocalizableString.cs

```diff
@@ -45,29 +43,61 @@ public override bool Equals(object other)
             return Equals(other as LocalizableString);
         }
 
-        private sealed class FixedLocalizableString : LocalizableString
+        internal LocalizableString MakeExceptionSafe()
         {
-            private readonly string _fixedString;
+            if (this is FixedLocalizableString || this is LocalizableResourceString)
+            {
+                // These are already sealed types which have exception safe implementations.
+                return this;
+            }
+
+            // Wrap the localizableString within an ExceptionSafeLocalizableString.
+            return new ExceptionSafeLocalizableString(this);
+        }
 
-            public FixedLocalizableString(string fixedResource)
+        void IExceptionSafeLocalizableString.SetOnException(Action<Exception> onException)
+        {
+            if (this is FixedLocalizableString)
             {
-                _fixedString = fixedResource;
+                // FixedLocalizableString can't throw.
+                return;
             }
 
-            public override string ToString(IFormatProvider formatProvider)
+            var localizableResourceString = this as LocalizableResourceString;
+            if (localizableResourceString != null)
             {
-                return _fixedString;
+                localizableResourceString.OnException = onException;
+                return;
             }
 
-            public override bool Equals(LocalizableString other)
+            // Must be a wrapped ExceptionSafeLocalizableString.
+            var exceptionSafeLocalizableString = (ExceptionSafeLocalizableString)this;
+            exceptionSafeLocalizableString.OnException = onException;
+        }
+
+        internal static T ExecuteAndCatchIfThrows<T>(Func<T> action, T defaulValueOnException, Action<Exception> onLocalizableStringException)
+        {
+            try
+            {
+                return action();
+            }
+            catch (Exception ex)
             {
-                var fixedStr = other as FixedLocalizableString;
-                return fixedStr != null && string.Equals(_fixedString, fixedStr.ToString());
+                onLocalizableStringException?.Invoke(ex);
+                return defaulValueOnException;
             }
+        }
 
-            public override int GetHashCode()
+        internal static T ExecuteAndCatchIfThrows<U, T>(Func<U, T> action, U argument, T defaulValueOnException, Action<Exception> onLocalizableStringException)
```

Comment:
don't you have this on some other thing as well? this should be just a generic helper thing.